url7315374482_toby1984_3dtest_tgz-pJ8-de_codesourcery_engine_Test3DJ8/src/de/codesourcery/engine/render/SoftwareRenderer.java

```diff
@@ -0,0 +1,720 @@
+			primitives.clear();
+		}
+
+		public void renderBatch(Graphics2D graphics) {
+
+		    if ( ! SKIP_RENDERING ) 
+		    {
+    			for ( PrimitiveWithDepth t : getTriangles() ) 
+    			{
+    				graphics.setColor( new Color( t.getColor() ) );
+    				drawPolygon( t.getPoints() , graphics , renderingMode );
+    			}
+		    }
+		    // TODO: Performance - Maybe clearing is not helping GC that much and should better be omitted ?
+			primitives.clear();
+		}
+
+		private List<PrimitiveWithDepth> getTriangles() 
+		{
+			Collections.sort( primitives , new Comparator<PrimitiveWithDepth>() {
+
+				@Override
+				public int compare(PrimitiveWithDepth o1, PrimitiveWithDepth o2) 
+				{
+					float z1 = o1.getDepth();
+					float z2 = o2.getDepth();
+
+					if ( z1 > z2 ) {
+						return -1;
+					} else if ( z1 < z2 ) {
+						return 1;
+					}
+					return 0;
+				}
+			});
+			return primitives;
+		}
+	}
+
+	public void setWorld(World world)
+	{
+		this.world = world;
+	}
+
+	protected final class RenderThread implements Runnable {
+
+		private final Graphics2D graphics;
+		private final Object3D object;
+		private final Matrix viewProjectionMatrix;
+
+		public RenderThread(Object3D object,Matrix viewProjectionMatrix,Graphics2D graphics) {
+			this.graphics = graphics;
+			this.object = object;
+			this.viewProjectionMatrix = viewProjectionMatrix;
+		}
+
+		@Override
+		public void run() 
+		{
+			final RenderingMode renderMode;
+			if ( object.isRenderOutline() ) {
+				renderMode = RenderingMode.RENDER_OUTLINE;
+			} else if ( object.isRenderWireframe() ) {
+				renderMode = RenderingMode.RENDER_WIREFRAME;
+			} else {
+				renderMode = RenderingMode.DEFAULT;
+			}
+
+			final PrimitiveBatch batch = new PrimitiveBatch(renderMode,object);
+
+			prepareRendering( object , viewProjectionMatrix , batch , graphics );
+			
+			if ( ! batch.isEmpty() ) {
+				batch.renderBatch( graphics );
+				
+			}
+		}
+	}
+
+	public void paint(final Graphics g)
+	{
+		final Graphics2D graphics = (Graphics2D) g;
+
+		// clear canvas
+		graphics.setColor(Color.LIGHT_GRAY);
+		graphics.fillRect( 0 , 0, getWidth() , getHeight() );
+
+		final List<Object3D> objects = world.getObjects();
+
+		final long start = -System.currentTimeMillis();
+
+		// ** rendering start **
+		
+		final Matrix viewProjectionMatrix = world.getProjectionMatrix().multiply(  world.getViewMatrix() );
+
+		// latch used to wait until all objects have been calculated
+		final CountDownLatch latch = new CountDownLatch( objects.size() );
+
+		final AtomicLong renderingTime = new AtomicLong(0); // updated by rendering thread
+		
+		synchronized( batches ) {
+			batches.clear();
+		}
+		
+		for( final Object3D obj : objects ) 
+		{
+			renderObject(graphics, viewProjectionMatrix, latch, renderingTime, obj , ! obj.hasChildren() );
+		}
+
+		// wait until primitives for all visible objects have been queued
+		while( true ) 
+		{
+			try {
+				if ( latch.await(5,TimeUnit.SECONDS) ) 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:** 
The loop containing this operation may iterate indefinitely if the operation throws an exception on multiple instances of its evaluation. As the operation throws an exception, it bypasses exit points in the loop and continues to the next iteration through a catch clause.

**Fix:** 
Add a retry counter to ensure that the loop terminates after a predefined number of iterations.